### PR TITLE
Enhance Transformation API

### PR DESCRIPTION
**Pull request by marcogschmidt**
This change adds new features to the transformation API. 
1. It is now possible to define `Extractor`s which target the request/response body and use the resulting data in templates. You can decide whether to parse the body as JSON or as raw text; in the latter case extraction happens via regular expressions, in the former you will also be able to use all the templating capabilities. 
2. The templating capabilities have been augmented with three new template functions: 
    - `body()` will return the whole request body,
    - `context()` will return the base JSON context (allowing for example to `range` on a JSON body that is an array),
    - `env(NAME)` will return the value of an environment variable with the given name, and finally
    - it is now possible to set [Dynamic Metadata](https://www.envoyproxy.io/docs/envoy/latest/configuration/advanced/well_known_dynamic_metadata) using all the available extraction and templating features. You can now use this metadata to include any request/response property you like in the access logs (see the `DYNAMIC_METADATA` command operator [here](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log#command-operators)).

For the actual changes to the API see [this commit](https://github.com/solo-io/gloo/pull/1547/commits/87c5fcaa85c36f8443252b9e6198442c2080bec4#diff-afe676ae7c02b5e1811021315a9756f8).

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/1546

### CODE REVIEW

File: projects/clusteringress/pkg/translator/translate_test.go [hunk 1]
```diff
@@ -230,7 +232,7 @@ var _ = Describe("Translate", func() {
 												},
 											},
 											RoutePlugins: &gloov1.RoutePlugins{
-												Transformations: (*transformation.RouteTransformations)(nil),
+												Transformations: (*envoytransformation.RouteTransformations)(nil),
 												Faults:          (*faultinjection.RouteFaults)(nil),
 												PrefixRewrite:   (*transformation.PrefixRewrite)(nil),
```
Comment (kdorosh) [inline projects/clusteringress/pkg/translator/translate_test.go]: can we just delete all instances of this? settings things to nil that are already nil by default offers no extra information

File: projects/gloo/api/external/envoy/extensions/transformation/transformation.proto [hunk 2]
```diff
@@ -70,6 +74,22 @@ message TransformationTemplate {
...
+    bool ignore_error_on_parse = 8;
+
+    message DynamicMetadataValue {
+        // optional. if not set filter namespace will be used.
+        string metadata_namespace = 1;
+        string key = 2 [(validate.rules).string = {min_bytes: 1}];
+        InjaTemplate value = 3;
+    }
+    repeated DynamicMetadataValue dynamic_metadata_values = 9;
+
```
[the remark this answers is not in the document]
Comment (marcogschmidt) [reply projects/gloo/api/external/envoy/extensions/transformation/transformation.proto]: Good point, will do.

File: projects/clusteringress/pkg/translator/translate_test.go (hunk 1 above)
Comment (marcogschmidt) [reply projects/clusteringress/pkg/translator/translate_test.go]: These were generated by the IDE when you do "Fill all fields" and there are 100+ occurences of those strings... :( Will see if I can find find a regex to remove them quickly

File: projects/clusteringress/pkg/translator/translate_test.go [hunk 3]
```diff
@@ -210,13 +203,13 @@ var _ = Describe("Translate", func() {
 																				XXX_sizecache:        0,
 																			},
 																		},
-																		DestinationSpec:      (*gloov1.DestinationSpec)(nil),
-																		Subset:               (*gloov1.Subset)(nil),
+																		DestinationSpec:      nil,
+																		Subset:               nil,
```
Comment (kdorosh) [inline projects/clusteringress/pkg/translator/translate_test.go]: we can do more than just delete the types. the default value is nil, so we should just delete this altogether

File: changelog/v0.20.12/new-transformation-features.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    and use the resulting data in templates. You can decide whether to parse the body as JSON or as raw text; in the
+    latter case extraction happens via regular expressions, in the former you will also be able to use all the templating
+    capabilities. These have been augmented with three new template functions: `body()` will return the whole request body,
+    `context()` will return the base JSON context (allowing for example to `range` on a JSON body that is an array), and
+    `env(NAME)` will return the value of an environment variable with the given name. Finally, it is now possible to set
+    [Dynamic Metadata](https://www.envoyproxy.io/docs/envoy/latest/configuration/advanced/well_known_dynamic_metadata)
+    using all the available extraction and templating features. You can now use this metadata to include any request/response
+    property you like in the access logs (see the `DYNAMIC_METADATA` command operator
+    [here](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log#command-operators)).
+  issueLink: https://github.com/solo-io/gloo/issues/1546
```
Comment (kdorosh) [inline changelog/v0.20.12/new-transformation-features.yaml]: `resolvesIssue`: `false` because we don't add support for accessing env vars

Comment (kdorosh) [reply changelog/v0.20.12/new-transformation-features.yaml]: ah its in the inja custom template 👍

**Pull request merged**